rclrs/src/wait/guard_condition.rs

```diff
@@ -0,0 +1,156 @@
+///
+/// // The wait call will now immediately return.
+/// ws.wait(Some(std::time::Duration::from_millis(10)))?;
+///
+/// # Ok::<(), RclrsError>(())
+/// ```
+pub struct GuardCondition {
+    /// The rcl_guard_condition_t that this struct encapsulates.
+    pub(crate) rcl_guard_condition: Arc<Mutex<rcl_guard_condition_t>>,
+    /// An optional callback to call when this guard condition is triggered.
+    callback: Option<Box<dyn Fn() + Send + Sync>>,
+    /// A flag to indicate if this guard condition has already been assigned to a wait set.
+    pub(crate) in_use_by_wait_set: Arc<AtomicBool>,
+}
+
+impl Drop for GuardCondition {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: No precondition for this function (besides passing in a valid guard condition)
+            rcl_guard_condition_fini(&mut *self.rcl_guard_condition.lock().unwrap());
+        }
+    }
+}
+
+// SAFETY: rcl_guard_condition is the only member that doesn't implement Send, and it is designed to be accessed from other threads
+unsafe impl Send for rcl_guard_condition_t {}
+
+// SAFETY: The mutexes and atomic members ensure synchronized access to members, and the callback is reentrant
+unsafe impl Sync for GuardCondition {}
+
+impl GuardCondition {
+    /// Creates a new guard condition.
+    pub fn new(context: &Context, callback: Option<Box<dyn Fn() + Send + Sync>>) -> Arc<Self> {
+        // SAFETY: Getting a zero initialized value is always safe
+        let mut guard_condition = unsafe { rcl_get_zero_initialized_guard_condition() };
+        unsafe {
+            // SAFETY: The context must be valid, and the guard condition must be zero-initialized
+            rcl_guard_condition_init(
+                &mut guard_condition,
+                &mut *context.rcl_context_mtx.lock().unwrap(),
+                rcl_guard_condition_get_default_options(),
+            );
+        }
+
+        Arc::new(Self {
+            rcl_guard_condition: Arc::new(Mutex::new(guard_condition)),
+            callback,
+            in_use_by_wait_set: Arc::new(AtomicBool::new(false)),
+        })
+    }
+
+    /// Triggers this guard condition, activating the wait set, and calling the optionally assigned callback.
+    pub fn trigger(&self) -> Result<(), RclrsError> {
+        unsafe {
+            // SAFETY: The rcl_guard_condition_t is valid.
+            rcl_trigger_guard_condition(&mut *self.rcl_guard_condition.lock().unwrap()).ok()?;
+        }
+        if let Some(callback) = &self.callback {
+            callback();
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::WaitSet;
+    use std::sync::atomic::Ordering;
+
+    #[test]
+    fn test_guard_condition() -> Result<(), RclrsError> {
+        let context = Context::new([])?;
+
+        let atomic_bool = Arc::new(std::sync::atomic::AtomicBool::new(false));
+        let atomic_bool_for_closure = Arc::clone(&atomic_bool);
+
+        let gc = GuardCondition::new(
```

Comment:
Minor nitpick, I prefer long names for variables (i.e. `guard_condition`), even if it's just tests.